[Lim Yu Zher] iP

LZher385 opened the pull request:
# Duke
>“I haven't been able to stop working ever since I used Duke!" - Yu Zher

Duke frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER FAST* to use

 All you need to do is,

1. download it from [here](https://github.com/LZher385/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is **FREE!**

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)
---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) {
        new Duke("./data/", "duke.txt").run();
    }
```

Commit:
Add Gradle support

LZher385 committed:
Add greetings, echo and exit

LZher385 committed:
Add ability to store and list user input

LZher385 committed:
Add the ability to mark tasks as done

LZher385 committed:
Add todos, events and deadlines as types of tasks

LZher385 committed:
Try testing using the I/O redirection technique

LZher385 committed:
Add error handling for some user inputs

LZher385 committed:
Add support for deleting tasks from the list

LZher385 committed:
Complete Level-7

LZher385 committed:
Complete Level-8

LZher385 committed:
Merge branch 'branch-Level-7'

LZher385 committed:
Merge branch 'branch-Level-8' and fix loading of file

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

LZher385 committed:
Make the code more OOP

LZher385 committed:
Organize into packages

LZher385 committed:
Add JUnit Tests

LZher385 committed:
Modify file path

LZher385 committed:
Create a JAR File

LZher385 committed:
Add JavaDoc comments

LZher385 committed:
Tweak the code to comply with the coding standard

LZher385 committed:
Add find functionality based on keyword

LZher385 committed:
Merge branch 'branch-A-JavaDoc'

LZher385 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/tasklist/TaskList.java

LZher385 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/tasklist/TaskList.java

mrmrinal reviewed (commented):
Hi Yu Zher, I have been assigned as the reviewer for your pull request to review your coding style.  Hope you found my comments useful!

mrmrinal commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,93 @@
+ * Contains methods which saves/deletes the list to a file.
+ */
+public class Storage {
+    private static final DateTimeFormatter FORMAT_TIME_FILE = DateTimeFormatter.ofPattern("MMM d yyyy hh:m a");
```
Good job in naming the constants using uppercase along with underscores!

mrmrinal commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,113 @@
+import duke.ui.Ui;
+
+import java.io.IOException;
+import java.time.DateTimeException;
```
I like how you spaced out your import statements based on the parent class it was imported from

LZher385 replied:
Thanks!

mrmrinal commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+     * @return A String representation of the duke.task.Deadline.
+     */
+    @Override
+    public String toString() {
```
Small nitpick, but overridden methods can get their javadocs copied from the parent class. The javadoc requires no substantial change unless the Overriden method does something completely different in the subclass.

LZher385 replied:
I see, that saves a whole bunch of effort as well. Thanks!

mrmrinal reviewed (commented):
Good Job. Your code followed the Coding Standards very closely, and errors were hard to spot.

LZher385 committed:
Add Gradle support

LZher385 committed:
Merge tag 'A-Gradle'

LZher385 committed:
Add JavaFX UI

LZher385 committed:
Merge branch 'branch-Level-10'

gok99 commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+     * @return Task with the specified index.
+     */
+    public String getTaskString(int index) { // RENAME
+        System.out.println(index + 1 + ": " + tasks.get(index));
```
repeated expression could be avoided with a variable

gok99 commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+     */
+    public String addDeadline(String fullDescription) throws DukeException {
+        int sepIndex = fullDescription.indexOf("/by");
+        if (fullDescription.charAt(sepIndex + 3) != ' ' || sepIndex == -1 || sepIndex == 0 ||
```
nit: this seems like a fairly complex expression that could be broken down and given meaningful names to help a little with code comprehension

gok99 commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+        String output = "";
+        ArrayList<Integer> taskIndexes = new ArrayList<>();
+        for (int i = 0; i < tasks.size(); i++) {
+            if (tasks.get(i).getDescription().toLowerCase(Locale.ROOT).contains(keyword.toLowerCase(Locale.ROOT))) {
```
another fairly complex expression that could benefit from being broken down

gok99 commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,193 @@
+        }
+        String output = "Noted. I have removed this task:\n";
+        output += "   " + tasks.get((taskId - 1)) + "\n";
+        System.out.println("Noted. I have removed this task:");
```
Console outputs could be abstracted out into a separate class, so that it would be easy to shift from CLI to GUI for example

gok99 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,95 @@
+                    int index = line.indexOf(" (at: ");
+                    String description = line.substring(10, index);
+                    String dateTime = line.substring(index + 6, line.length() - 1);
+                    String[] dateTimeArray = dateTime.split(" ");
```
there seems to be a lot of repeated code here, perhaps some abstraction here would help keep things DRY

gok99 commented on src/main/java/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
+Main-Class: duke.Duke
+
```
not sure if this file is meant to be committed

gok99 reviewed (commented):
Nice work! Code quality seems to me to be mostly ok, except for a few issues that I could spot :)

zacharylwy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,93 @@
+import java.io.IOException;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Scanner;
```
OO pretty nice organisation for the import statements. Will learn from you!

zacharylwy commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+     * @param description Description of the task.
+     * @param dateString  Date and time of the deadline.
+     * @param formatInput How the input is formatted.
+     * @param hasTime     Boolean for whether time was supplied.
```
Nice readability for the comments!

zacharylwy commented on build.gradle:
```diff
@@ -0,0 +1,46 @@
+    toolVersion = '8.29'
+}
+
+run{
```
{ should have a spacing!

LZher385 committed:
Add assertions

LZher385 committed:
Improve code quality

There are a few sections of repeated codes and long expressions.

Modifying them makes things easier to maintain in the long run.

Let's
* Abstract out the repeated code into functions.

LZher385 committed:
Merge pull request #1 from LZher385/branch-A-Assertions

Add assertions

LZher385 committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/tasklist/TaskList.java

LZher385 committed:
Merge pull request #2 from LZher385/branch-A-CodeQuality

Improve code quality

LZher385 committed:
Add snooze functionality

Allow users to snooze tasks by a set number of days.

This provides greater flexibility for task dates.

Let's
* provide a command for specifying task to modify and number of
days to increase by
* ensure the changes are saved locally

LZher385 committed:
Merge branch 'branch-B-Snooze'

LZher385 committed:
Refactor method length

LZher385 committed:
Modify Java comments

LZher385 committed:
Adjust padding of text in GUI

LZher385 committed:
Change UI images

LZher385 committed:
Modify wrapping of lines in GUI

LZher385 committed:
Add user guide and UI image

LZher385 committed:
Set theme jekyll-theme-slate

LZher385 committed:
Crop UI image to fit requirements

LZher385 committed:
Merge branch 'master' of https://github.com/LZher385/ip
